pipeline-runner/R/handle_data.R

```diff
@@ -51,17 +42,64 @@ reload_scdata_from_s3 <- function(pipeline_config, experiment_id, task_name, tas
   obj <- readRDS(rawConnection(body))
   return(obj)
 }
+reload_scdata_list_from_s3 <- function (s3, experiment_id) {
+  bucket <- pipeline_config$source_bucket
+  objects <- s3$list_objects(
+    Bucket = bucket,
+    Prefix = experiment_id
+  )
+  samples <- objects$Contents
+
+  scdata_list <- list()
+  for (sample in samples) {
+    key <- sample$Key
+
+    c(body, ...rest) %<-% s3$get_object(
+      Bucket = bucket,
+      Key = paste(key, sep = "/")
+    )
+    obj <- readRDS(rawConnection(body))
+    sample_id <- strsplit(key, "/")[[1]][[2]]
+    scdata_list[[sample_id]] <- obj
+  }
+
+  return(scdata_list)
+}
+
+# reload_data_from_s3 will reload:
+# * scdata_list for all steps before integration (included)
+# * scdata file for all steps after data integration
+reload_data_from_s3 <- function(pipeline_config, experiment_id, task_name, tasks) {
```

Comment:
this function returns either an rds object (if the task is after data integration and function `reload_scdata_from_s3` runs) or a list of objects (if function `reload_scdata_list_from_s3` runs). How does the code upstream manage to handle those two possible types of output that `reload_data_from_s3` can return?

Reply:
The functions downstream either accept a list (everything before integration) or a merged seurat object (everything after integration).